tests/codegen/vec-in-place.rs

```diff
@@ -36,40 +36,60 @@ pub struct Baz {
 // CHECK-LABEL: @vec_iterator_cast_primitive
 #[no_mangle]
 pub fn vec_iterator_cast_primitive(vec: Vec<i8>) -> Vec<u8> {
+    // FIXME(jieyouxu): these matches are complete hacks. Please find a way to write them in a less
+    // insane manner...
+
     // CHECK-NOT: loop
-    // CHECK-NOT: call
+    // CHECK-NOT: call void @{{.*}}Iterator{{.*}}collect
+    // CHECK-NOT: call void @{{.*}}Iterator{{.*}}map
```

Comment:
This is very not nice, but I also don't know of a better way to make this less hacky.

Reply:
What does the IR look like now? On current master this optimizes down to basically nothing, just copying the vec from the argument to the return slot:
```ll
define void @vec_iterator_cast_primitive(ptr noalias nocapture noundef writeonly sret([24 x i8]) align 8 dereferenceable(24) %_0, ptr noalias nocapture noundef readonly align 8 dereferenceable(24) %vec) unnamed_addr #0 personality ptr @rust_eh_personality {
start:
  %me.sroa.0.0.copyload.i = load i64, ptr %vec, align 8, !alias.scope !3, !noalias !6
  %me.sroa.4.0.self.sroa_idx.i = getelementptr inbounds i8, ptr %vec, i64 8
  %me.sroa.4.0.copyload.i = load ptr, ptr %me.sroa.4.0.self.sroa_idx.i, align 8, !alias.scope !3, !noalias !6, !nonnull !8, !noundef !8
  %me.sroa.5.0.self.sroa_idx.i = getelementptr inbounds i8, ptr %vec, i64 16
  %me.sroa.5.0.copyload.i = load i64, ptr %me.sroa.5.0.self.sroa_idx.i, align 8, !alias.scope !3, !noalias !6
  store i64 %me.sroa.0.0.copyload.i, ptr %_0, align 8, !alias.scope !9, !noalias !12
  %vec.sroa.4.0._0.sroa_idx.i = getelementptr inbounds i8, ptr %_0, i64 8
  store ptr %me.sroa.4.0.copyload.i, ptr %vec.sroa.4.0._0.sroa_idx.i, align 8, !alias.scope !9, !noalias !12
  %vec.sroa.5.0._0.sroa_idx.i = getelementptr inbounds i8, ptr %_0, i64 16
  store i64 %me.sroa.5.0.copyload.i, ptr %vec.sroa.5.0._0.sroa_idx.i, align 8, !alias.scope !9, !noalias !12
  ret void
}
```

Reply:
It's more like

```llvm
; Function Attrs: uwtable
define void @vec_iterator_cast_primitive(ptr noalias nocapture noundef writeonly sret([24 x i8]) align 8 dereferenceable(24) %_0, ptr noalias nocapture noundef readonly align 8 dereferenceable(24) %vec) unnamed_addr #0 personality ptr @rust_eh_personality {
start:
  %_9.i.i.i.i.i = alloca %"core::option::Option<core::fmt::Arguments<'_>>", align 8
  %_6.i.i.i.i.i = alloca %"core::option::Option<usize>", align 8
  %upper.i.i.i.i.i = alloca %"core::option::Option<usize>", align 8
  %_9.i.i.i.i = alloca %"core::option::Option<core::fmt::Arguments<'_>>", align 8
  %_6.i.i.i.i = alloca %"core::option::Option<usize>", align 8
  %upper.i.i.i.i = alloca %"core::option::Option<usize>", align 8
  %self = alloca %"core::iter::adapters::map::Map<alloc::vec::into_iter::IntoIter<i8>, {closure@/Users/jieyouxu/Documents/Repos/rust/tests/codegen/vec-in-place.rs:45:25: 45:28}>", align 8
  %me.sroa.0.0.copyload.i = load i64, ptr %vec, align 8, !alias.scope !286, !noalias !289
  %me.sroa.4.0.self.sroa_idx.i = getelementptr inbounds i8, ptr %vec, i64 8
  %me.sroa.4.0.copyload.i = load ptr, ptr %me.sroa.4.0.self.sroa_idx.i, align 8, !alias.scope !286, !noalias !289, !nonnull !2, !noundef !2
  %me.sroa.5.0.self.sroa_idx.i = getelementptr inbounds i8, ptr %vec, i64 16
  %me.sroa.5.0.copyload.i = load i64, ptr %me.sroa.5.0.self.sroa_idx.i, align 8, !alias.scope !286, !noalias !289
  %_19.i = getelementptr inbounds i8, ptr %me.sroa.4.0.copyload.i, i64 %me.sroa.5.0.copyload.i
  store ptr %me.sroa.4.0.copyload.i, ptr %self, align 8
  %self1.sroa.4.0.self.sroa_idx = getelementptr inbounds i8, ptr %self, i64 8
  store ptr %me.sroa.4.0.copyload.i, ptr %self1.sroa.4.0.self.sroa_idx, align 8
  %self1.sroa.5.0.self.sroa_idx = getelementptr inbounds i8, ptr %self, i64 16
  store i64 %me.sroa.0.0.copyload.i, ptr %self1.sroa.5.0.self.sroa_idx, align 8
  %self1.sroa.6.0.self.sroa_idx = getelementptr inbounds i8, ptr %self, i64 24
  store ptr %_19.i, ptr %self1.sroa.6.0.self.sroa_idx, align 8
  tail call void @llvm.experimental.noalias.scope.decl(metadata !291)
  call void @llvm.lifetime.start.p0(i64 16, ptr nonnull %upper.i.i.i.i), !noalias !294
  store i8 1, ptr %upper.i.i.i.i, align 8, !noalias !294
  %0 = getelementptr inbounds i8, ptr %upper.i.i.i.i, i64 8
  store i64 %me.sroa.5.0.copyload.i, ptr %0, align 8, !noalias !294
  call void @llvm.lifetime.start.p0(i64 16, ptr nonnull %_6.i.i.i.i), !noalias !294
  %1 = getelementptr inbounds i8, ptr %_6.i.i.i.i, i64 8
  store i64 %me.sroa.5.0.copyload.i, ptr %1, align 8, !noalias !294
  store i64 1, ptr %_6.i.i.i.i, align 8, !noalias !294
  %2 = load i64, ptr %upper.i.i.i.i, align 8, !range !12, !alias.scope !302, !noalias !305, !noundef !2
  %trunc.not.i.i.i.i.i = icmp eq i64 %2, 0
  br i1 %trunc.not.i.i.i.i.i, label %bb3.i.i.i.i, label %"_ZN5alloc3vec16in_place_collect108_$LT$impl$u20$alloc..vec..spec_from_iter..SpecFromIter$LT$T$C$I$GT$$u20$for$u20$alloc..vec..Vec$LT$T$GT$$GT$9from_iter17h855d9ac797569b5aE.exit"
```
now (with yet more basic blocks)

Reply:
The purpose of that test is to ensure pretty much what erikdesjardins is quoting there, i.e. that the whole operation is a noop and the remaining IR is just passing through the struct. So if this method were inlined into a caller it would be a complete noop.

>   %_9.i.i.i.i = alloca %"core::option::Option<core::fmt::Arguments<'_>>", align 8

Semantically this method does not do any formatting at all, so that sounds like some panic path isn't optimized out. Which means that it is indeed failing the purpose of the test.

Reply:
> Semantically this method does not do any formatting at all, so that sounds like some panic path isn't optimized out. Which means that it is indeed failing the purpose of the test.

I posted my comment just before I fixed the critical bug with `from_immediate_scalar`, which also happens to fix this. Now it no longer contains any calls so I can just undo my changes to that test:

```rs
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) uwtable
define void @vec_iterator_cast_primitive(ptr noalias nocapture noundef writeonly sret([24 x i8]) align 8 dereferenceable(24) %_0, ptr noalias nocapture noundef readonly align 8 dereferenceable(24) %vec) unnamed_addr #0 personality ptr @rust_eh_personality {
start:
  %me.sroa.0.0.copyload.i = load i64, ptr %vec, align 8, !alias.scope !2, !noalias !5
  %me.sroa.4.0.self.sroa_idx.i = getelementptr inbounds i8, ptr %vec, i64 8
  %me.sroa.4.0.copyload.i = load ptr, ptr %me.sroa.4.0.self.sroa_idx.i, align 8, !alias.scope !2, !noalias !5, !nonnull !7, !noundef !7
  %me.sroa.5.0.self.sroa_idx.i = getelementptr inbounds i8, ptr %vec, i64 16
  %me.sroa.5.0.copyload.i = load i64, ptr %me.sroa.5.0.self.sroa_idx.i, align 8, !alias.scope !2, !noalias !5
  store i64 %me.sroa.0.0.copyload.i, ptr %_0, align 8, !alias.scope !8, !noalias !11
  %vec.sroa.4.0._0.sroa_idx.i = getelementptr inbounds i8, ptr %_0, i64 8
  store ptr %me.sroa.4.0.copyload.i, ptr %vec.sroa.4.0._0.sroa_idx.i, align 8, !alias.scope !8, !noalias !11
  %vec.sroa.5.0._0.sroa_idx.i = getelementptr inbounds i8, ptr %_0, i64 16
  store i64 %me.sroa.5.0.copyload.i, ptr %vec.sroa.5.0._0.sroa_idx.i, align 8, !alias.scope !8, !noalias !11
  ret void
}
```

Reply:
Hmm, that looks like a regression, seems like it doesn't get optimized out anymore.

We'll probably have to figure out why this is happening (and whether we need e.g. more LLVM optimizations) before we can merge.

Reply:
> Hmm, that looks like a regression, seems like it doesn't get optimized out anymore.

(Yes, it was a regression when my ABI was completely broken and caused the optimization to not work, but now after not having ABI problems it no longer fails the test in its unmodified form)